diff -Nur OpenSim-/Framework/WebUtil.cs OpenSim/Framework/WebUtil.cs
--- OpenSim-/Framework/WebUtil.cs	2025-04-06 19:53:58.816200196 +0900
+++ OpenSim/Framework/WebUtil.cs	2025-04-23 14:52:47.364090084 +0900
@@ -1423,6 +1423,9 @@
                 responseMessage = client.Send(request, HttpCompletionOption.ResponseHeadersRead);
                 responseMessage.EnsureSuccessStatusCode();

+                // Instructions for synchronization by Fumi.Hax
+                string dummy = responseMessage.Content.ReadAsStringAsync().Result;
+
                 int rcvlen = 0;
                 if ((responseMessage.Content.Headers.ContentLength is long contentLength) && contentLength != 0)
                 {
@@ -1517,6 +1520,9 @@
                 responseMessage = client.Send(request, HttpCompletionOption.ResponseHeadersRead);
                 responseMessage.EnsureSuccessStatusCode();

+                // Instructions for synchronization by Fumi.Hax
+                string dummy = responseMessage.Content.ReadAsStringAsync().Result;
+
                 int rcvlen = 0;
                 if ((responseMessage.Content.Headers.ContentLength is long contentLength) && contentLength != 0)
                 {
diff -Nur OpenSim-/Region/CoreModules/World/Land/LandManagementModule.cs OpenSim/Region/CoreModules/World/Land/LandManagementModule.cs
--- OpenSim-/Region/CoreModules/World/Land/LandManagementModule.cs	2025-04-06 19:53:58.859200692 +0900
+++ OpenSim/Region/CoreModules/World/Land/LandManagementModule.cs	2025-04-23 14:52:47.362090060 +0900
@@ -530,6 +530,10 @@
             if (avatar.IsChildAgent || avatar.IsNPC)
                 return;
 
+            // by Fumi.Hax
+            ILandObject over = GetLandObject(avatar.AbsolutePosition.X, avatar.AbsolutePosition.Y);
+            SendLandUpdate(avatar, over);
+
             if (m_showBansLines && !m_scene.RegionInfo.EstateSettings.TaxFree)
                 SendOutNearestBanLine(avatar.ControllingClient);
         }
diff -Nur OpenSim-/Region/OptionalModules/Avatar/XmlRpcGroups/XmlRpcGroupsServicesConnectorModule.cs OpenSim/Region/OptionalModules/Avatar/XmlRpcGroups/XmlRpcGroupsServicesConnectorModule.cs
--- OpenSim-/Region/OptionalModules/Avatar/XmlRpcGroups/XmlRpcGroupsServicesConnectorModule.cs	2025-04-06 19:53:58.899201153 +0900
+++ OpenSim/Region/OptionalModules/Avatar/XmlRpcGroups/XmlRpcGroupsServicesConnectorModule.cs	2025-04-23 14:52:47.363090072 +0900
@@ -1154,7 +1154,8 @@
 
             using (Stream stream = request.GetRequestStream())
             {
-                using (XmlTextWriter xml = new XmlTextWriter(stream, Encoding.ASCII))
+                //using (XmlTextWriter xml = new XmlTextWriter(stream, Encoding.ASCII))
+                using (XmlTextWriter xml = new XmlTextWriter(stream, Encoding.UTF8))    // by Fumi.Hax
                 {
                     _serializer.Serialize(xml, this);
                     xml.Flush();
diff -Nur OpenSim-/Region/OptionalModules/UserStatistics/WebStatsModule.cs OpenSim/Region/OptionalModules/UserStatistics/WebStatsModule.cs
--- OpenSim-/Region/OptionalModules/UserStatistics/WebStatsModule.cs	2025-04-06 19:53:58.905201222 +0900
+++ OpenSim/Region/OptionalModules/UserStatistics/WebStatsModule.cs	2025-04-23 14:55:34.059057627 +0900
@@ -143,7 +143,9 @@
                 m_scenes.Add(scene);
                 updateLogMod = m_scenes.Count * 2;
 
-                m_simstatsCounters.Add(scene.RegionInfo.RegionID, new USimStatsData(scene.RegionInfo.RegionID));
+                // by Fumi.Hax
+                if (!m_simstatsCounters.ContainsKey(scene.RegionInfo.RegionID))
+                    m_simstatsCounters.Add(scene.RegionInfo.RegionID, new USimStatsData(scene.RegionInfo.RegionID));
 
                 scene.EventManager.OnRegisterCaps += OnRegisterCaps;
                 scene.EventManager.OnDeregisterCaps += OnDeRegisterCaps;
diff -Nur OpenSim-/Server/Handlers/Freeswitch/FreeswitchServerConnector.cs OpenSim/Server/Handlers/Freeswitch/FreeswitchServerConnector.cs
--- OpenSim-/Server/Handlers/Freeswitch/FreeswitchServerConnector.cs	2025-04-06 19:53:58.950201741 +0900
+++ OpenSim/Server/Handlers/Freeswitch/FreeswitchServerConnector.cs	2025-04-23 14:56:24.106648349 +0900
@@ -105,7 +105,8 @@
                 if (s.Trim() != "")
                 {
                     string [] nvp = s.Split(new Char [] {'='});
-                    bodyParams.Add(HttpUtility.UrlDecode(nvp[0]), HttpUtility.UrlDecode(nvp[1]));
+                    if (!bodyParams.ContainsKey(nvp[0]))  // by Fumi.Hax
+                        bodyParams.Add(HttpUtility.UrlDecode(nvp[0]), HttpUtility.UrlDecode(nvp[1]));
                 }
             }
 
